aptos/contracts/sources/pyth.move

```diff
@@ -0,0 +1,803 @@
+        };
+
+        sources
+    }
+
+    #[test_only]
+    /// Expose a public initialization function for use in tests
+    public fun init_test(
+        signer_capability: account::SignerCapability,
+        stale_price_threshold: u64,
+        governance_emitter_chain_id: u64,
+        governance_emitter_address: vector<u8>,
+        data_sources: vector<DataSource>,
+        update_fee: u64,
+    ) {
+        init_internal(
+            signer_capability,
+            stale_price_threshold,
+            governance_emitter_chain_id,
+            governance_emitter_address,
+            data_sources,
+            update_fee
+        )
+    }
+
+// -----------------------------------------------------------------------------
+// Update the cached prices
+
+    /// Update the cached price feeds with the data in the given VAAs. This is a 
+    /// convenience wrapper around update_price_feeds(), which allows you to update the price feeds
+    /// using an entry function.
+    /// 
+    /// If possible, it is recommended to use update_price_feeds() instead, which avoids the need
+    /// to pass a signer account. update_price_feeds_with_funder() should only be used when
+    /// you need to call an entry function.
+    /// 
+    /// This function will charge an update fee, transferring some AptosCoin's
+    /// from the given funder account to the Pyth contract. The amount of coins transferred can be
+    /// queried with get_update_fee(). The signer must have sufficient account balance to
+    /// pay this fee, otherwise the transaction will abort.
+    public entry fun update_price_feeds_with_funder(account: &signer, vaas: vector<vector<u8>>) {
+        let coins = coin::withdraw<AptosCoin>(account, get_update_fee());
+        update_price_feeds(vaas, coins);
+    }
+
+    /// Update the cached price feeds with the data in the given VAAs.
+    /// The vaas argument is a vector of VAAs encoded as bytes.
+    /// 
+    /// The given fee must contain a sufficient number of coins to pay the update fee.
+    /// The update fee amount can be queried by calling get_update_fee().
+    public fun update_price_feeds(vaas: vector<vector<u8>>, fee: Coin<AptosCoin>) {
+        // Update the price feed from each VAA
+        while (!vector::is_empty(&vaas)) {
+            update_price_feed_from_single_vaa(vector::pop_back(&mut vaas));
+        };
+
+        // Charge the message update fee
+        assert!(get_update_fee() <= coin::value(&fee), error::insufficient_fee());
+        coin::deposit(@pyth, fee);
+    }
+
+    fun update_price_feed_from_single_vaa(vaa: vector<u8>) {
+        // Deserialize the VAA
+        let vaa = vaa::parse_and_verify(vaa);
+
+        // Check that the VAA is from a valid data source (emitter)
+        assert!(
+            state::is_valid_data_source(
+                data_source::new(
+                    u16::to_u64(vaa::get_emitter_chain(&vaa)),
+                    vaa::get_emitter_address(&vaa))),
```

Comment:
I think it's better to get the update_fee here rather than the provided additional fee (which might be given by mistake)